File: j-backend/src/routes/v1/customers.js
/**
 * Customers API routes.
 */
const express = require('express');
const asyncHandler = require('express-async-handler');
const { v4: uuidv4 } = require('uuid');
const { authenticate, requireRole } = require('../../middleware/auth');
const { Customer, Sale } = require('../../models');
const { Op } = require('sequelize');

const router = express.Router();

// POST / - Create customer
router.post('', authenticate, asyncHandler(async (req, res) => {
  if (!req.user.shop_id) {
    return res.status(400).json({ detail: 'User is not associated with a shop' });
  }
  
  const existing = await Customer.findOne({
    where: { shop_id: req.user.shop_id, phone: req.body.phone },
  });
  
  if (existing) {
    return res.status(400).json({ detail: 'Customer with this phone number already exists' });
  }
  
  const newCustomer = await Customer.create({
    id: uuidv4(),
    shop_id: req.user.shop_id,
    ...req.body,
  });
  
  res.status(201).json(newCustomer);
}));

// GET / - List customers
router.get('', authenticate, asyncHandler(async (req, res) => {
  if (!req.user.shop_id) {
    return res.json({ customers: [], total: 0, skip: 0, limit: 100 });
  }
  
  const { skip = 0, limit = 100, search, is_active } = req.query;
  
  const where = { shop_id: req.user.shop_id };
  
  if (search) {
    where[Op.or] = [
      { name: { [Op.iLike]: `%${search}%` } },
      { phone: { [Op.iLike]: `%${search}%` } },
      { email: { [Op.iLike]: `%${search}%` } },
    ];
  }
  
  if (is_active !== undefined) {
    where.is_active = is_active === 'true';
  }
  
  const { count, rows } = await Customer.findAndCountAll({
    where,
    order: [['name', 'ASC']],
    offset: parseInt(skip, 10),
    limit: parseInt(limit, 10),
  });
  
  res.json({
    customers: rows,
    total: count,
    skip: parseInt(skip, 10),
    limit: parseInt(limit, 10),
  });
}));

// GET /:customer_id
router.get('/:customer_id', authenticate, asyncHandler(async (req, res) => {
  const customer = await Customer.findOne({
    where: { id: req.params.customer_id, shop_id: req.user.shop_id },
  });
  
  if (!customer) {
    return res.status(404).json({ detail: 'Customer not found' });
  }
  
  res.json(customer);
}));

// PATCH /:customer_id
router.patch('/:customer_id', authenticate, asyncHandler(async (req, res) => {
  const customer = await Customer.findOne({
    where: { id: req.params.customer_id, shop_id: req.user.shop_id },
  });
  
  if (!customer) {
    return res.status(404).json({ detail: 'Customer not found' });
  }
  
  await customer.update(req.body);
  
  res.json(customer);
}));

// DELETE /:customer_id
router.delete('/:customer_id', authenticate, requireRole(['owner', 'manager', 'admin']), asyncHandler(async (req, res) => {
  const customer = await Customer.findOne({
    where: { id: req.params.customer_id, shop_id: req.user.shop_id },
  });
  
  if (!customer) {
    return res.status(404).json({ detail: 'Customer not found' });
  }
  
  // Soft delete
  await customer.update({ is_active: false });
  
  res.status(204).send();
}));

// GET /:customer_id/purchase-history
router.get('/:customer_id/purchase-history', authenticate, asyncHandler(async (req, res) => {
  const customer = await Customer.findOne({
    where: { id: req.params.customer_id, shop_id: req.user.shop_id },
  });
  
  if (!customer) {
    return res.status(404).json({ detail: 'Customer not found' });
  }
  
  const { skip = 0, limit = 50 } = req.query;
  
  const sales = await Sale.findAll({
    where: { shop_id: req.user.shop_id },
    order: [['created_at', 'DESC']],
  });
  
  // Filter by customer phone
  const customerSales = sales.filter(
    s => s.customer_info?.phone === customer.phone,
  );
  
  const paginatedSales = customerSales.slice(
    parseInt(skip, 10),
    parseInt(skip, 10) + parseInt(limit, 10),
  );
  
  res.json({
    customer_id: customer.id,
    customer_name: customer.name,
    total_purchases: parseFloat(customer.total_purchases),
    total_orders: customerSales.length,
    purchases: paginatedSales.map(s => ({
      sale_id: s.id,
      invoice_no: s.invoice_no,
      date: s.created_at.toISOString(),
      total_amount: parseFloat(s.total_amount),
      payment_type: s.payment_type,
      status: s.status,
    })),
    skip: parseInt(skip, 10),
    limit: parseInt(limit, 10),
    total: customerSales.length,
  });
}));

module.exports = router;

File: backend/app/api/v1/customers.py
"""
Customer management API endpoints.
"""
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from uuid import UUID
from datetime import datetime
import uuid

from app.db.session import get_db
from app.models.user import User
from app.models.customer import Customer
from app.utils.auth import get_current_active_user, require_role
from app.schemas.customer import (
    CustomerCreate,
    CustomerUpdate,
    CustomerResponse,
    CustomerListResponse
)

router = APIRouter()


@router.post("", response_model=CustomerResponse, status_code=status.HTTP_201_CREATED)
async def create_customer(
    customer_data: CustomerCreate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """
    Create a new customer.
    
    Args:
        customer_data: Customer data
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Created customer
    """
    if not current_user.shop_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="User is not associated with a shop"
        )
    
    # Check for duplicate phone in same shop
    existing = db.query(Customer).filter(
        Customer.shop_id == current_user.shop_id,
        Customer.phone == customer_data.phone
    ).first()
    
    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Customer with this phone number already exists"
        )
    
    new_customer = Customer(
        id=uuid.uuid4(),
        shop_id=current_user.shop_id,
        **customer_data.model_dump()
    )
    
    db.add(new_customer)
    db.commit()
    db.refresh(new_customer)
    
    return new_customer


@router.get("", response_model=CustomerListResponse)
async def list_customers(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    search: str | None = Query(None),
    is_active: bool | None = Query(None),
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """
    List all customers for the current shop.
    
    Args:
        skip: Number of records to skip
        limit: Maximum number of records to return
        search: Search by name, phone, or email
        is_active: Filter by active status
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        List of customers
    """
    if not current_user.shop_id:
        return CustomerListResponse(customers=[], total=0, skip=skip, limit=limit)
    
    query = db.query(Customer).filter(Customer.shop_id == current_user.shop_id)
    
    if search:
        search_pattern = f"%{search}%"
        query = query.filter(
            (Customer.name.ilike(search_pattern)) |
            (Customer.phone.ilike(search_pattern)) |
            (Customer.email.ilike(search_pattern))
        )
    
    if is_active is not None:
        query = query.filter(Customer.is_active == is_active)
    
    total = query.count()
    customers = query.order_by(Customer.name).offset(skip).limit(limit).all()
    
    return CustomerListResponse(
        customers=customers,
        total=total,
        skip=skip,
        limit=limit
    )


@router.get("/{customer_id}", response_model=CustomerResponse)
async def get_customer(
    customer_id: UUID,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """
    Get a specific customer by ID.
    
    Args:
        customer_id: Customer UUID
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Customer details
    """
    customer = db.query(Customer).filter(
        Customer.id == customer_id,
        Customer.shop_id == current_user.shop_id
    ).first()
    
    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found"
        )
    
    return customer


@router.patch("/{customer_id}", response_model=CustomerResponse)
async def update_customer(
    customer_id: UUID,
    customer_data: CustomerUpdate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """
    Update a customer.
    
    Args:
        customer_id: Customer UUID
        customer_data: Updated customer data
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Updated customer
    """
    customer = db.query(Customer).filter(
        Customer.id == customer_id,
        Customer.shop_id == current_user.shop_id
    ).first()
    
    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found"
        )
    
    # Update fields
    update_data = customer_data.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(customer, field, value)
    
    customer.updated_at = datetime.utcnow()
    db.commit()
    db.refresh(customer)
    
    return customer


@router.delete("/{customer_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_customer(
    customer_id: UUID,
    current_user: User = Depends(require_role(["owner", "manager", "admin"])),
    db: Session = Depends(get_db)
):
    """
    Delete a customer (soft delete by setting is_active=False).
    Only owners, managers, and admins can delete customers.
    
    Args:
        customer_id: Customer UUID
        current_user: Current authenticated user
        db: Database session
    """
    customer = db.query(Customer).filter(
        Customer.id == customer_id,
        Customer.shop_id == current_user.shop_id
    ).first()
    
    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found"
        )
    
    # Soft delete
    customer.is_active = False
    customer.updated_at = datetime.utcnow()
    db.commit()
    
    return None


@router.get("/{customer_id}/purchase-history")
async def get_customer_purchase_history(
    customer_id: UUID,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=500),
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """
    Get purchase history for a customer.
    
    Args:
        customer_id: Customer UUID
        skip: Number of records to skip
        limit: Maximum number of records to return
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Customer's purchase history
    """
    from app.models.sales import Sale
    
    customer = db.query(Customer).filter(
        Customer.id == customer_id,
        Customer.shop_id == current_user.shop_id
    ).first()
    
    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found"
        )
    
    # Find sales with this customer's phone in customer_info
    sales = db.query(Sale).filter(
        Sale.shop_id == current_user.shop_id
    ).all()
    
    # Filter by customer phone (stored in customer_info JSON)
    customer_sales = [
        s for s in sales 
        if s.customer_info.get('phone') == customer.phone
    ]
    
    # Apply pagination
    total = len(customer_sales)
    customer_sales = customer_sales[skip:skip+limit]
    
    return {
        "customer_id": str(customer_id),
        "customer_name": customer.name,
        "total_purchases": float(customer.total_purchases),
        "total_orders": len(customer_sales),
        "purchases": [
            {
                "sale_id": str(s.id),
                "invoice_no": s.invoice_no,
                "date": s.created_at.isoformat(),
                "total_amount": float(s.total_amount),
                "payment_type": s.payment_type,
                "status": s.status
            }
            for s in customer_sales
        ],
        "skip": skip,
        "limit": limit,
        "total": total
    }

